### ISSUE TITLE
First pass at a smoke test harness for repo extraction. [phab:D2293]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2757]
_Created by **danielmoy-google** at 2018-05-09 20:41:07_: 

There's open issues in this commit which should be addressed in future
commits:

1. currently relies on freebie quota, instead this needs to fix x_oauth2
import and call with a proper github token.

2. directory paths don't match up. needs some improvement to do more
fuzzy matching (be tolerant to the first few path elements not matching
up)

3. reading from github shouldn't be blocked by failed extraction -
should consider doing in parallel or something.  Alternatively, leave
this as-is because github quota exhaustion is a problem - might as well
not read from github individually if the clone fails

4. split apart, not have all logic in one giant file

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-09 23:41:25_:

This is now broken as of cody's latest changes.  Not quite sure yet what I need to do to fix it, will look tomorrow.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-10 00:04:26_:

  - Merge branch 'master' into tester

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-10 00:06:01_:

merged back from master and fixed resulting build problems

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 16:58:09_:

Consider including the config flag in the usage doc here. Also, consider adding a link to the proto defining the config for reference.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 16:58:09_:

This line will never be executed, is it required by the compiler?

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 16:58:09_:

The wording in the usage string seems a bit awkward, consider: "A file which contains a newline delimited list of repos to test."

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2018-05-10 18:25:12_:

#2761 uses Gazelle to import all third_party Go dependencies.  This change should be updated to also do so.

**Comment by kythe** [comment]
_**schroederc** required changes at 2018-05-10 18:25:12_

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:23_:

Would it be more efficient to default to recursively reading a tree in order to reduce the total number of round trips per repo? The response for the request includes a "truncated" field which could be used as cue to handle the case when the 200 file limit is reached: https://developer.github.com/v3/git/trees/#get-a-tree

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:23_:

Is there a reason we're hard-coding the restriction to java files only? It seems like this approach could be creating more work for us later. Perhaps a more broad pattern could be used incorporating the suffixes of languages currently supported by Kythe. Perhaps this pattern could be defined in an accessible/prominent constant or config for ease of modification with the enabling of new languages. Or alternatively, this could be passed in as a command line flag.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:23_:

Same comment as line 296

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:23_:

Is this left over debugging code?

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-05-10 20:45:23_:

One general comment, there don't seem to be any tests for the logic of this program. Relying on the github API complicates testing, however one way to address this is implement an interface for the github interactions which you can shim for the tests.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:24_:

An approach to consider which could potentially avoid the rate limit being hit, is to look into utilizing the Pagination option the go github API provides: https://github.com/google/go-github#pagination

This would require some sort of bath request end-point in the Github API, as well as implemented support within the go library, either of which may not exist, but it could be worth looking in to.

In lieu of that, some sort of parallel fan-out and fan-in could reduce the total runtime in cases with large repo lists.

**Comment by kythe** [comment]
_**craigbarber** commented inline (context lost during migration) at 2018-05-10 20:45:24_:

If this turns out to be a bottle-neck for large repos, there could be an opportunity for parallelism here as well, as individiual sub-trees can be crawled in parallel, followed by a barrier then merge the results prior to returning.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-11 18:51:25_:

  - update so that github/querystring work.
  - Addressing comments.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

I am unclear how to proceed here given shahms' rollback of #2761.

I was partway through making this change but now a merge with master puts me back to the original version here.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

Ah yea I was partway through refactoring this and forgot.

Technically the nil is required here in this current snapshot, yes.

Though also I didn't want to leave it like this, so thanks for pointing it out anyways :)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

yea definitely, though I won't tackle that until after auth, rate limiting, etc.  Added a comment as a reminder.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

so unfortunately I think this doesn't work.

My parse of go-github API could be wrong, but my understanding is that the pagination option only works for API which specify opt as a paramemter.  GetTree is not one such API.

The good news is that GetTree does support a recursive option.

The bad news is that it's just a boolean toggle, where the additional failover case just yields "go back and do this non-recursively" as the only workaround.  So that's not so ideal.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

I'm not sure.  For sufficiently small repos maybe.  I guess one we do testing I'll collect some metrics on how many repos contain <200 total files (where directory counts as a file), and see if that is enough coverage to warrant additional logic in here.

My sneaking suspicion is that porting to use githubql instead will be the better way:

https://developer.github.com/v4/

I'm not settled either way yet.

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

yea definitely, I imagine this being passed as a configurable option in the future, along side selection of what extractor(s) to run.

I'd imagine this might also come with some better way of packaging up the extractor bits (rather than relying on just previously running 'blaze build kythe/java/com/google/devtools/kythe/extractors/java/artifact' ahead of time)

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2018-05-11 18:52:07_:

ah yea, good catch thanks

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-11 18:52:07_:

Thanks for the comments :)

Note some of them are probably out of scope for this first check-in (auth, parallelization, pagination/recursive, etc), but I'm noting them in the bug as open work items (and some of them are left as TODO also).

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-05-11 19:44:55_:

>>! In (earlier comment), craigbarber wrote:
> One general comment, there don't seem to be any tests for the logic of this program. Relying on the github API complicates testing, however one way to address this is implement an interface for the github interactions which you can shim for the tests.

This comment may have been lost in the noise. Wanted to ping it to make sure it's considered :)

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-11 19:53:42_:

>>! In (earlier comment), craigbarber wrote:
>>>! In (earlier comment), craigbarber wrote:
>> One general comment, there don't seem to be any tests for the logic of this program. Relying on the github API complicates testing, however one way to address this is implement an interface for the github interactions which you can shim for the tests.
> 
> This comment may have been lost in the noise. Wanted to ping it to make sure it's considered :)

Yea this I'm not done with.  Also not sure whether this commit should be significantly modified to have an accompanying test, or if I should do that as a follow-up refactor.

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-11 19:57:12_:

>>! In (earlier comment), danielmoy-google wrote:
>>>! In (earlier comment), craigbarber wrote:
>> This comment may have been lost in the noise. Wanted to ping it to make sure it's considered :)
> 
> Yea this I'm not done with.  Also not sure whether this commit should be significantly modified to have an accompanying test, or if I should do that as a follow-up refactor.

In particular I'd guess we'd want to judiciously split this up into library stuff.

I'd sort of rather get in a commit with working 3p deps, based on the amount of time I've now burned juggling between different formats of 3p go library importing since cody's changing all of this stuff as I go.

**Comment by kythe** [comment]
_**craigbarber** wrote at 2018-05-11 20:02:51_:

Makes sense, just want to make sure it doesn't get lost "like tear drops in rain" :)

**Comment by kythe** [comment]
_**craigbarber** accepted this revision at 2018-05-11 20:02:51_

**Comment by kythe** [comment]
_**danielmoy-google** wrote at 2018-05-11 20:04:25_:

>>! In (earlier comment), craigbarber wrote:
> Makes sense, just want to make sure it doesn't get lost "like tear drops in rain" :)

We can (probably should) promote refactor ahead of other items for future improvement (except maybe auth)

**Comment by kythe** [comment]
_**danielmoy-google** closed via commit 86c5d43084570ae3d49d6595dd6f6054bf9a39d9 at 2018-05-11 20:06:12_
